source/Account.cpp

```diff
@@ -150,322 +219,448 @@ void Account::PayExtra(int mortgage, int64_t amount)
+	//    1. The number of credits paid towards ALL fines
+	//    2. Whether ANY fine was not paid in full
+	Receipt receipt;
+	if(mortgages.empty()) return receipt;
 
-	// If you made payments of three or more types, the punctuation needs to
-	// include commas, so just handle that separately here.
-	if(typesPaid.size() >= 3)
+	for(Mortgage &mortgage : mortgages)
 	{
-		auto it = typesPaid.begin();
-		for(unsigned int i = 0; i < typesPaid.size() - 1; ++i)
+		if(!(mortgage.Type() == "Fine")) continue;
+		int64_t payment = mortgage.Payment();
+		if(payment > credits)
 		{
-			out << Format::CreditString(it->second) << " in " << it->first << ", ";
-			++it;
+			mortgage.MissPayment();
+			receipt.paidInFull = false;
+		}
+		else
+		{
+			payment = mortgage.MakePayment();
+			credits -= payment;
+			receipt.creditsPaid += payment;
 		}
-		out << "and " << Format::CreditString(it->second) << " in " << it->first + ".";
-	}
-	else
-	{
-		if(salariesPaid)
-			out << Format::CreditString(salariesPaid) << " in crew salaries"
-				<< ((mortgagesPaid || finesPaid || debtPaid || maintenancePaid) ? " and " : ".");
-		if(maintenancePaid)
-			out << Format::CreditString(maintenancePaid) << "  in maintenance"
-				<< ((mortgagesPaid || finesPaid || debtPaid) ? " and " : ".");
-		if(mortgagesPaid)
-			out << Format::CreditString(mortgagesPaid) << " in mortgages"
-				<< ((finesPaid || debtPaid) ? " and " : ".");
-		if(finesPaid)
-			out << Format::CreditString(finesPaid) << " in fines"
-				<< (debtPaid ? " and " : ".");
-		if(debtPaid)
-			out << Format::CreditString(debtPaid) << " in debt.";
 	}
-	return out.str();
-}
-
 
-
-const map<string, int64_t> &Account::SalariesIncome() const
-{
-	return salariesIncome;
+	return receipt;
 }
 
 
 
-int64_t Account::SalariesIncomeTotal() const
+Receipt Account::PayDebts()
 {
-	return accumulate(
-		salariesIncome.begin(),
-		salariesIncome.end(),
-		0,
-		[](int64_t value, const std::map<string, int64_t>::value_type &salary)
+	// This function needs to preserve:
+	//    1. The number of credits paid towards ALL debts
+	//    2. Whether ANY fine was not paid in full
+	Receipt receipt;
+	if(mortgages.empty()) return receipt;
+
+	for(Mortgage &mortgage : mortgages)
+	{
+		if(!(mortgage.Type() == "Debt")) continue;
+		int64_t payment = mortgage.Payment();
+		if(payment > credits)
 		{
-			return value + salary.second;
+			mortgage.MissPayment();
+			receipt.paidInFull = false;
 		}
-	);
-}
-
-
+		else
+		{
+			payment = mortgage.MakePayment();
+			credits -= payment;
+			receipt.creditsPaid += payment;
+		}
+	}
 
-void Account::SetSalaryIncome(const string &name, int64_t amount)
-{
-	if(amount == 0)
-		salariesIncome.erase(name);
-	else
-		salariesIncome[name] = amount;
+	return receipt;
 }
 
 
 
-int64_t Account::CrewSalariesOwed() const
+const string Account::GenerateMissedPaymentLogs(std::vector<Receipt> *receipts) const
 {
-	return crewSalariesOwed;
-}
+	ostringstream log;
 
+	if(!receipts->at(0).paidInFull)
```

Comment:
Using the index in the vector as a magic number to determine which item we are looking at is not very elegant and maintainable.

Maybe consider putting in the receipt itself what the receipt is for? (So the receipt at position 0 would have a string in it that says "crew salaries", and the receipt at position 1 would have a string that says "maintenance costs".)

Reply:
No, documenting the magic numbers would not be sufficient. And I'm not looking for complex solutions with maps or tuple headers.
It's just that we have information hidden within the positions of the vector, and that is makes the code less maintainable.

Reply:
I don't disagree, magic numbers are something to be avoided most of the time. There's a couple of ways to do what you'd like as I see it:
1. Create a map of Receipt and type (salary, maintenance, mortgages), which requires linking the map headers and defeats the original purpose of using the custom struct, or
2. Add a type enum to the struct, then loop through the receipt vector and check for the type to operate on, which adds an unnecessary loop and is very amateurish, despite the performance impact from a loop that small and run once per time step is functionally zero.

I'm happy to go with option 1, but I do think this is one of the rare cases where magic numbers are inoffensive. Thoughts?

Reply:
Both options 1 and 2 don't sound great, and I don't know if my suggestion from July 10th to add a string to the Receipt struct itself (so that the Receipt would internally have the information on what it is for) would work.

Magic numbers are still not okay here, because they seem to make the code quite a bit harder to read.

Please try to find a solution that is easier to read (and thus more maintainable).